In [3]:
import pandas

In [2]:
mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


In [2]:
import pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2],
   'travel':['1km','23km','7km']
}

myvar = pd.DataFrame(mydataset)

print(myvar)

    cars  passings travel
0    BMW         3    1km
1  Volvo         7   23km
2   Ford         2    7km


In [8]:
# To Find The pandas version
import pandas as pd

print(pd.__version__)

1.5.3


# Pandas Series

What is a Series?

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [10]:
# a = [1, 7, 2]
a = [1, 7, 2, 5, 5,4 ,5]
myvar = pd.Series(a)

print(myvar)

0    1
1    7
2    2
3    5
4    5
5    4
6    5
dtype: int64


In [11]:
# Labels
print(myvar[0]) # Return the first value of the Series


1


In [31]:
# Create Labels

a = [1, 7, 2]
b=[[1,2],[1,2],[1,2]]
myvar = pd.Series(a, index = ["x", "y", "z"])
myvar1 = pd.Series(b, index = ["x", "y", "z"])
print(myvar)
print(myvar1)

x    1
y    7
z    2
dtype: int64
x    [1, 2]
y    [1, 2]
z    [1, 2]
dtype: object


In [38]:
print(myvar["y"]) # Now it return the value of label

7


In [36]:
s = pd.Series([1,2,3,4,5],index = ['a','b','c','d','e'])

#retrieve multiple elements
print(s[['a','c','d']])

a    1
c    3
d    4
dtype: int64


In [26]:
# Key/Value Objects as Series

# Create a simple Pandas Series from a dictionary:

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)


day1    420
day2    380
day3    390
dtype: int64


In [23]:
# Create a Series using only data from "day1" and "day2":

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)


day1    420
day2    380
dtype: int64


In [24]:
# DataFrames

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)


   calories  duration
0       420        50
1       380        40
2       390        45


# Pandas DataFrames

What is a DataFrame? <br>
A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.


In [39]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 

   calories  duration
0       420        50
1       380        40
2       390        45


In [40]:
# Locate Row
'''As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the loc attribute to return one or more specified row(s)'''

#refer to the row index:
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


In [41]:
# Return row 0 and 1:

#use a list of indexes:
print(df.loc[[0, 1]])

   calories  duration
0       420        50
1       380        40


In [46]:
# Named Indexes

# With the index argument, you can name your own indexes.

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df) 

      calories  duration
day1       420        50
day2       380        40
day3       390        45


In [47]:
# Return "day2":

#refer to the named index:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


In [ ]:
# Load Files Into a DataFrame

# If your data sets are stored in a file, Pandas can load them into a DataFrame.

In [49]:
df = pd.read_csv('data.csv')
print(df) # Print the DataFrame without the to_string() method:

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


# Pandas Read CSV
A simple way to store big data sets is to use CSV files (comma separated files).
<br>
CSV files contains plain text and is a well know format that can be read by everyone including Pandas.
<br>
In our examples we will be using a CSV file called 'data.csv'.

[Download data.csv](https://www.w3schools.com/python/pandas/data.csv). or [Open data.csv](https://www.w3schools.com/python/pandas/data.csv.txt)


In [48]:
df = pd.read_csv('data.csv')

print(df.to_string()) # Tip: use to_string() to print the entire DataFrame.

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

If you have a large DataFrame with many rows, Pandas will only return the first 5 rows, and the last 5 rows:
## max_rows

The number of rows returned is defined in Pandas option settings.
<br>
You can check your system's maximum rows with the pd.options.display.max_rows statement.


In [50]:
print(pd.options.display.max_rows) 


60


In my system the number is 60, which means that if the DataFrame contains more than 60 rows, the print(df) statement will return only the headers and the first and last 5 rows.
<br>
You can change the maximum rows number with the same statement.

In [51]:
# Increase the maximum number of rows to display the entire DataFrame:

pd.options.display.max_rows = 9999

df = pd.read_csv('data.csv')

print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

# Read JSON
Big data sets are often stored, or extracted as JSON.
<br>
JSON is plain text, but has the format of an object, and is well known in the world of programming, including Pandas.
<br>
In our examples we will be using a JSON file called 'data.json'.
<br>
[Open data.json](https://www.w3schools.com/python/pandas/data.js).

In [4]:
# Load the JSON file into a DataFrame:

a = pd.read_json("data.json")
print(a)
print(a.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]
     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147   

### Dictionary as JSON

In [6]:
'''JSON = Python Dictionary

JSON objects have the same format as Python dictionaries.'''

# Load a Python Dictionary into a DataFrame:
#Neat Example

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df) 

    Duration  Pulse  Maxpulse  Calories
0       60.0  110.0     130.0     409.0
1       60.0    NaN     145.0     479.0
2       60.0  103.0     135.0     340.0
3       45.0  109.0     175.0     282.0
4       45.0  117.0     148.0     406.0
5       60.0  102.0     127.0     300.0
12       NaN  117.0       NaN       NaN


In [7]:
# Error Example Comare the above code and this code
'''JSON = Python Dictionary

JSON objects have the same format as Python dictionaries.'''

# Load a Python Dictionary into a DataFrame:

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "12":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df) 

    Duration  Pulse  Maxpulse  Calories
0       60.0  110.0     130.0     409.0
1       60.0    NaN     145.0     479.0
2       60.0  103.0     135.0     340.0
3       45.0  109.0     175.0     282.0
4       45.0  117.0     148.0     406.0
5       60.0  102.0     127.0     300.0
12       NaN  117.0       NaN       NaN


# Pandas - Analyzing DataFrames

# # Viewing the Data
One of the most used method for getting a quick overview of the DataFrame, is the head() method.
<br>
The head() method returns the headers and a specified number of rows, starting from the top.

In [8]:
df = pd.read_csv('data.csv')

print(df.head(10))
# Note: if the number of rows is not specified, the head() method will return the top 5 rows.

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


There is also a tail() method for viewing the last rows of the DataFrame.
<br>
The tail() method returns the headers and a specified number of rows, starting from the bottom.

In [11]:
# Print the last 5 rows of the DataFrame:
print(df.tail()) 

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


### Info About the Data
The DataFrames object has a method called info(), that gives you more information about the data set.


In [12]:
# Print information about the data:
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


### Result Explained
The result tells us there are 169 rows and 4 columns:
```
  RangeIndex: 169 entries, 0 to 168
  Data columns (total 4 columns):
```
And the name of each column, with the data type:

```
   #   Column    Non-Null Count  Dtype  
  ---  ------    --------------  -----  
   0   Duration  169 non-null    int64  
   1   Pulse     169 non-null    int64  
   2   Maxpulse  169 non-null    int64  
   3   Calories  164 non-null    float64
```

